src/main/java/software/amazon/cloudwatchlogs/emf/model/Metric.java

```diff
@@ -21,65 +21,66 @@
+    protected String name;
 
     @Getter
     @JsonProperty("Unit")
     @JsonSerialize(using = UnitSerializer.class)
     @JsonDeserialize(using = UnitDeserializer.class)
-    private Unit unit;
+    protected Unit unit;
 
     @Getter
-    @Setter
     @JsonProperty("StorageResolution")
     @JsonInclude(
             value = JsonInclude.Include.CUSTOM,
             valueFilter =
                     StorageResolutionFilter.class) // Do not serialize when valueFilter is true
     @JsonSerialize(using = StorageResolutionSerializer.class)
-    public StorageResolution storageResolution;
+    protected StorageResolution storageResolution;
 
-    @JsonIgnore @NonNull @Getter private List<Double> values;
+    @JsonIgnore @Getter protected Object values;
 
-    MetricDefinition(String name) {
-        this(name, Unit.NONE, StorageResolution.STANDARD, new ArrayList<>());
+    /**
+     * Change the name of this metric. Should only be used within this package in MetricContext when
+     * recieving an unnamed metric from a user.
+     *
+     * @param name The new metric name.
+     */
+    protected void setName(String name) {
+        if (name == null) {
+            throw new NullPointerException("Metric name cannot be null");
+        }
+        this.name = name;
     }
 
-    MetricDefinition(String name, double value) {
-        this(name, Unit.NONE, StorageResolution.STANDARD, value);
-    }
-
-    MetricDefinition(String name, Unit unit, double value) {
-        this(name, unit, StorageResolution.STANDARD, new ArrayList<>(Arrays.asList(value)));
-    }
+    /**
+     * Creates a Metric with the first {@code size} values of the current metric
+     *
+     * @param size the maximim size of the returned metric's values
+     * @return a Metric with the first {@code size} values of the current metric.
+     */
+    protected abstract Metric getMetricValuesUnderSize(int size);
 
-    MetricDefinition(String name, StorageResolution storageResolution, double value) {
-        this(name, Unit.NONE, storageResolution, new ArrayList<>(Arrays.asList(value)));
-    }
+    /**
+     * Creates a Metric all metrics after the first {@code size} values of the current metric. If
+     * there are less than {@code size} values, null is returned.
+     *
+     * @param size the maximim size of the returned metric's values
+     * @return a Metric with the all metrics after the first {@code size} values of the current
+     *     metric. If there are less than {@code size} values, null is returned.
+     */
+    protected abstract Metric getMetricValuesOverSize(int size);
 
-    MetricDefinition(String name, Unit unit, StorageResolution storageResolution, double value) {
-        this(name, unit, storageResolution, new ArrayList<>(Arrays.asList(value)));
-    }
-
-    void addValue(double value) {
-        values.add(value);
-    }
+    /** @return a simplified representation of the values of this metric. */
+    protected abstract Object getSimplifiedValues();
```

Comment:
qq: How would this look like for histogram/stats?

Reply:
Statistics cannot be simplified so it would be the same as `getValues()`. For histograms if there are only 1 or 2 distinct values they would each be their own bucket as per the SEH1 algorithm implementation I am following.

Reply:
I think I'll rename this function to `getPublishingValues()` or something similar, to imply that it is getting values in the format that they will be published in.

Reply:
Updated name to `getFormattedValues()`